discord.go

```diff
@@ -0,0 +1,272 @@
+
+	discordImpl.session, err = discordgo.New("Bot " + discordImpl.botToken)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare Discord session: %s", err)
+	}
+	discordImpl.logger.Info("opening websocket connection to Discord")
+
+	discordImpl.session.Open()
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to open websocket connection to Discord: %s", err)
+	}
+
+	discordImpl.logger.Infof("successfully opened websocket connection to Discord")
+
+	discordImpl.logger.Infof("attaching slash command handler")
+
+	// c.Logger.Infof("reloading roles from config")
+	// err := c.reloadRoles()
+	// if err != nil {
+	// 	return fmt.Errorf("failed to reload config roles")
+	// }
+
+	err = discordImpl.registerSlashCommands()
+	if err != nil {
+		return nil, fmt.Errorf("failed to register slash commands: %s", err)
+	}
+
+	message, err := discordImpl.SendChannelMessage(discordImpl.primaryChannelID, "Hello! Thank you for inviting me!")
+	if err != nil {
+		return nil, fmt.Errorf("failed to send welcome message: %s", message.Content)
+	}
+	discordImpl.logger.Infof("successfully sent welcome message: %s", message.Content)
+
+	return discordImpl, nil
+}
+
+func (Discord *DiscordImpl) registerSlashCommands() error {
+	Discord.logger.Infof("registering slash commands")
+	var def bool = false
+	// Create an array of pointers to discordgo.ApplicationCommand structs
+	commands := []*discordgo.ApplicationCommand{
+		{
+			Name:              "panicalert",
+			Description:       "Start an alert admin vote.",
+			DefaultPermission: &def,
+			Options: []*discordgo.ApplicationCommandOption{
+				{
+					Type:        discordgo.ApplicationCommandOptionString,
+					Name:        "message",
+					Description: "The message to send to the admin.",
+					Required:    true,
+				},
+			},
+		},
+		{
+			Name:              "panicban",
+			Description:       "The user whom the ban vote is about.",
+			DefaultPermission: &def,
+			Options: []*discordgo.ApplicationCommandOption{
+				{
+					Type:        discordgo.ApplicationCommandOptionString,
+					Name:        "reason",
+					Description: "Reason why this user should be banned.",
+					Required:    true,
+				},
+				{
+					Type:        discordgo.ApplicationCommandOptionInteger,
+					Name:        "days",
+					Description: "The number of days of previous messages to delete",
+					Required:    false,
+				},
+			},
+		},
+	}
+	// map the names of the commands to their callback
+	commandHandlers := map[string]func(s *discordgo.Session, i *discordgo.InteractionCreate){
+		"panicalert":    Discord.panicAlertCallback,
+		"panicban":      Discord.panicBanCallback,
+		"embedReaction": Discord.embedReactionCallback,
+	}
+
+	// Add a listener for when the Discord API fires an InteractionCreate event.
+	Discord.session.AddHandler(func(s *discordgo.Session, i *discordgo.InteractionCreate) {
+		if handler, ok := commandHandlers[i.ApplicationCommandData().Name]; ok {
+			handler(s, i)
```

Comment:
I don't think this will work for the embed reaction but I may be wrong.

Reply:
I agree with you. I think what this code is doing is it is checking the ApplicationCommandData.Name field available on i which is a *discordgo.InteractionCreate. I am checking the commandHandlers slice for a name that matches one of the keys inside the slice.

Since we will not have a slash command named embedReaction this will not work for when an "embedReaction event" is fired.
However this should work for panicalert and panicban because there are slash commands that are registered with those names. ok would be true and the handler would be called to handle those events.